File: src/linked/double/DoubleNode.ts
import {Node} from "../Node";
export default class DoubleNode extends Node{
    child?:DoubleNode;
    parent?:DoubleNode;
    constructor(val:number,public parentNode?:DoubleNode,childNode?:DoubleNode) {
        super(val,childNode);
        this.parent = parentNode;
    }

    insert(val: number): DoubleNode | null {
        if(!this.child){
            this.child = new DoubleNode(val,this);
            return this.child;
        }
        return this.child.insert(val);
    }
    remove(val: number): DoubleNode | null {
        if(!this.child) return null;
        if(this.child.value === val){
            this.child = <DoubleNode>this.child.child;
            if(this.child){
                this.child.parent = this;
            }else{
                return this;
            }
        }
        return this.child.remove(val);
    }
}

File: linkedlist/DoubleNode.java
package linkedlist;
public class DoubleNode extends Node{
    private INode parent;
    public DoubleNode(int value,INode parent) {
        super(value);
        this.parent = parent;
    }
    public DoubleNode(int value,INode parent,INode child) {
        super(value,child);
        this.parent = parent;
    }
    public INode getParent(){
        return this.parent;
    }
    public void setParent(INode parent){
        this.parent = parent;
    }
    public void delete(int val){
        if(this.child == null) return;
        if(this.child.getValue() == val){
            INode grandChildNode = this.child.getChild();
            if(grandChildNode != null){
                ((DoubleNode)grandChildNode).parent = this;
            }
            this.child = grandChildNode;
            return;
        }
        ((DoubleNode)this.child).delete(val);
    }

}